src/tyr/route_serializer_osrm.cc

```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
   return geojson;
 }
 
+// Maps Valhalla edge and road class to an OpenLR form-of-way.
+FormOfWay RoadClassToFOW(const TripLeg::Node& node) {
+  const bool is_missing_rc = !node.has_edge() || !node.edge().has_road_class();
+  if (is_missing_rc || !node.edge().has_use()) {
+    return FormOfWay::OTHER;
+  }
+  const TripLeg::Use use = node.edge().use();
+  const valhalla::RoadClass rc = node.edge().road_class();
+  const bool one_way =
+      node.edge().has_traversability() && node.edge().traversability() != TripLeg::kBoth;
+  if (node.edge().roundabout()) {
+    return FormOfWay::ROUNDABOUT;
+  } else if (use == TripLeg::kRampUse || use == TripLeg::kTurnChannelUse) {
+    return FormOfWay::SLIPROAD;
+  } else if (rc == valhalla::RoadClass::kMotorway && one_way) {
+    return FormOfWay::MOTORWAY;
+  } else if (rc <= valhalla::RoadClass::kTertiary && one_way) {
+    return FormOfWay::MULTIPLE_CARRIAGEWAY;
+  } else if (rc <= valhalla::RoadClass::kTertiary) {
+    return FormOfWay::SINGLE_CARRIAGEWAY;
+  } else {
+    return FormOfWay::OTHER;
+  }
+}
+
+// TODO(mookerji): Move this shared header file somewhere and remove also from tests/openlr.cc
+std::string encode64(const std::string& val) {
+  using namespace boost::archive::iterators;
+  using It = base64_from_binary<transform_width<std::string::const_iterator, 6, 8>>;
+  auto tmp = std::string(It(std::begin(val)), It(std::end(val)));
+  return tmp.append((3 - val.size() % 3) % 3, '=');
+}
+
+// TODO(mookerji): comment
+std::string openlr_shape(const valhalla::TripRoute& route) {
+  std::vector<OpenLR::LocationReferencePoint> lrps;
+  for (const TripLeg& leg : route.legs()) {
+    const std::vector<PointLL>& points = midgard::decode<std::vector<midgard::PointLL>>(leg.shape());
+    float bearing_deg = 0;
+    float distance_meters = 0;
+    for (std::size_t i = 0, num_nodes = leg.node_size(); i < num_nodes; ++i) {
+      const TripLeg::Node& current = leg.node(i);
+      const PointLL& location = points.at(current.edge().end_shape_index());
+      const FormOfWay fow = RoadClassToFOW(current);
+      const unsigned char frc = static_cast<unsigned char>(fow);
+      if (i == num_nodes - 1) {
+        bearing_deg = fmod(bearing_deg + 180., 360.);
+        lrps.emplace_back(location.lng(), location.lat(), bearing_deg, frc, fow,
+                          lrps.empty() ? nullptr : &lrps.back());
+      } else {
+        const PointLL& next_location = points.at(leg.node(i + 1).edge().end_shape_index());
+        bearing_deg = location.Heading(next_location);
+        distance_meters = location.Distance(next_location);
+      }
+      // TODO(continued): should the start node for a new leg point to the lrps of the last node of
```

Comment:
yeah i think that because you smashing all the legs together it would make more sense to only add the last lrp of a leg if it is the last leg (and also only then flip the bearing). basically add a break to the top of the nodes for loop that is like

```c++
if(&leg == &route.legs().back())
  break;
```